### PR TITLE
chore: compare micro benchmarks

### PR DESCRIPTION
**Pull request by alankritdabral**
**Summary:**  
Compare Criterion benchmarks with the cache_benchmarks 
Added instructions to compare the benchmarks locally
run:
`rust-script benches/criterion_compare <benchmarks1.json> <benchmarks2.json> check` to check the degradation in benchmarks
`rust-script benches/criterion_compare <benchmarks1.json> <benchmarks2.json> table` to get the comparison table
`rust-script benches/criterion_compare <benchmarks1.json> <benchmarks2.json> ` to get the comparison table and then check the degradation.

**Issue Reference(s):**  
Fixes #... _(Replace "..." with the issue number)_

**Build & Testing:**

- [ ] I ran `cargo test` successfully.
- [ ] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation website] accordingly.
- [ ] I have performed a self-review of my own code.

[documentation website]: https://github.com/tailcallhq/tailcallhq.github.io

### CODE REVIEW

File: .github/workflows/bot.yml [hunk 1]
```diff
@@ -183,3 +183,57 @@ jobs:
           key: criterion_benchmarks_${{ github.sha }}
           restore-keys: |
             criterion_benchmarks_
+
+  Criterion_compare:
+    name: Comparing Micro Benchmarks
+    if: contains(github.event.pull_request.labels.*.name, 'benchmark') || github.event_name == 'push'
+    runs-on: ubuntu-latest
```
Comment (amitksingh1490) [inline .github/workflows/bot.yml:190]:
```suggestion
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
      contents: write
```

**Comment by tusharmath** [comment]
- [ ] Benchmarks should always run and results should be cached.
- [ ] Comparison should only happen on PRs
- [ ] CI should fail if the benchmarks show a performance degradation of greater than 1%

**Comment by alankritdabral** [comment]
i have two jobs `Cache_Benchmarks` `Criterion_compare`

`Cache_Benchmarks`:

1. switch to main branch 
2.  run benchmarks 
3.  write to cache 
4. comment the benchmarks

`Criterion_compare`

1. read the cache 
2. run the benchmark on current 
3. comment the current benchmarks on pr
4. comment the comparison on the CLI

File: .github/workflows/bot.yml [hunk 2]
```diff
@@ -160,9 +160,12 @@ jobs:
   Cache_Benchmarks:
     name: Micro Benchmarks
     runs-on: ubuntu-latest
+    permissions:
+      pull-requests: write
+      contents: write
     steps:
       - name: Check out code
-        uses: actions/checkout@v4
+        uses: actions/checkout@v2
```
Comment (amitksingh1490) [inline .github/workflows/bot.yml:168]:
```suggestion
        uses: actions/checkout@v4
```

File: .github/contributing.md [hunk 3]
```diff
@@ -47,6 +47,29 @@ Thank you for considering contributing to **Tailcall**! This document outlines t
...
+   ```bash
+   git checkout main
+   cargo criterion --message-format=json > benches/main_benchmarks.json
+   git checkout -
+   cargo criterion --message-format=json > benches/benchmarks.json
+   rust-script benches/criterion_compare benches/main_benchmarks.json benches/benchmarks.json
+
+   ```
+
+4. **Check the Results:** If the benchmarks show more than 10% degradation, the script will exit with an error. Please check "benches/benchmark.md" file to identify the benchmarks that failed and investigate the code changes that might have caused the degradation.
```
Comment (amitksingh1490) [inline .github/contributing.md:71]:
```suggestion
4. **Check the Results:** If the benchmarks show more than 1% degradation, the script will exit with an error. Please check "benches/benchmark.md" file to identify the benchmarks that failed and investigate the code changes that might have caused the degradation.
```

File: .github/workflows/bot.yml [hunk 4]
```diff
@@ -183,3 +187,42 @@ jobs:
...
+          toolchain: stable
+          override: true
+
+      - name: Restore file
+        uses: actions/cache@v2
+        with:
+          path: benches/main_benchmarks.json
+          key: criterion_benchmarks_${{ github.sha }}
+          restore-keys: |
+            criterion_benchmarks_
```
Comment (tusharmath) [inline .github/workflows/bot.yml:214]: Why do we need this?

File: .github/workflows/bot.yml [hunk 5]
```diff
@@ -183,3 +187,42 @@ jobs:
...
+
+      - name: Print Criterion Benchmarks
+        run: |
+          cargo install cargo-criterion rust-script
+          cargo criterion --message-format=json > benches/benchmarks.json
+          rust-script benches/parse_json_benchmark benches/benchmarks.json benches/change_benchmarks.md
+          cat benches/change_benchmarks.md
```
Comment (tusharmath) [inline .github/workflows/bot.yml:221]: - Keep the `cat` step to "Print Benchmarks"
- Create a new "Run Benchmarks" step that performs the running of the benchmarks and creation of the md file.

File: .github/workflows/bot.yml (hunk 4 above)
Comment (alankritdabral) [reply .github/workflows/bot.yml:214]: it is used to restore the cache file (main benchmarks in json )

File: .github/workflows/bot.yml [hunk 6]
```diff
@@ -183,3 +187,43 @@ jobs:
           key: criterion_benchmarks_${{ github.sha }}
           restore-keys: |
             criterion_benchmarks_
+
+  Criterion_compare:
```
Comment (tusharmath) [inline .github/workflows/bot.yml:191]:
```suggestion
  Criterion_Compare:
```

File: .github/workflows/bot.yml [hunk 7]
```diff
@@ -174,7 +177,8 @@ jobs:
         run: |
           cargo install cargo-criterion rust-script
           cargo criterion --message-format=json > benches/main_benchmarks.json
-          rust-script benches/parse_json_benchmark benches/main_benchmarks.json
+          rust-script benches/parse_json_benchmark benches/main_benchmarks.json benches/main_benchmarks.md
```
Comment (tusharmath) [inline .github/workflows/bot.yml:180]:
```suggestion
          rust-script benches/json_to_md benches/main_benchmarks.json benches/main_benchmarks.md
```

File: .github/workflows/bot.yml [hunk 8]
```diff
@@ -160,6 +160,9 @@ jobs:
   Cache_Benchmarks:
     name: Micro Benchmarks
     runs-on: ubuntu-latest
+    permissions:
+      pull-requests: write
+      contents: write
```
Comment (tusharmath) [inline .github/workflows/bot.yml:165]: Run this only on main

File: .github/workflows/bot.yml [hunk 9]
```diff
@@ -183,3 +187,43 @@ jobs:
...
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          override: true
+
+      - name: Restore file
+        uses: actions/cache@v3
+        with:
+          path: benches/main_benchmarks.json
+          key: criterion_benchmarks_${{ github.sha }}
```
Comment (alankritdabral) [inline .github/workflows/bot.yml:212]: github.event.pull_request.base.sha

File: benches/criterion_compare.rs [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+  }
+}
+
+fn calculate_percentage_change(old_value: f64, new_value: f64) -> f64 {
+  if old_value == 0.0 {
+    0.0
+  } else {
+    ((new_value - old_value) / old_value) * 100.0
+  }
+}
```
Comment (tusharmath) [inline benches/criterion_compare.rs:167]: move this file to `/scripts`

File: benches/json_to_md.rs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+  // Attempt to parse the JSON file and print the table
+  match parse_json(input_file_path) {
+    Ok(_) => println!("Table printed successfully."),
+    Err(err) => {
+      eprintln!("Error: {}", err);
+      std::process::exit(1);
+    }
+  }
+}
```
Comment (tusharmath) [inline benches/json_to_md.rs:127]: move this file to `/scripts`

File: .github/workflows/bot.yml [hunk 12]
```diff
@@ -156,10 +178,50 @@ jobs:
...
+        with:
+          path: benches/main_benchmarks.json
+          key: criterion_benchmarks_${{ github.event.pull_request.base.sha }}
+          fail-on-cache-miss: true
+
+      - name: Run Criterion Benchmarks
+        run: |
+          cargo install cargo-criterion rust-script
+          cargo criterion --message-format=json > benches/benchmarks.json
+          ./scripts/json_to_md benches/benchmarks.json > benches/change_benchmarks.md
```
Comment (alankritdabral) [inline .github/workflows/bot.yml:218]: ./scripts/json_to_md.rs benches/benchmarks.json > benches/change_benchmarks.md

File: .github/workflows/bot.yml [hunk 13]
```diff
@@ -53,6 +53,28 @@ jobs:
...
+          sudo cp nginx.conf /etc/nginx/
+
+      - name: Test Nginx Configuration
+        run: |
+          sudo nginx -t
+
+      - name: Start Nginx
+        run: |
+          sudo systemctl start nginx
+
```
Comment (tusharmath) [inline .github/workflows/bot.yml:77]: Nginx changes here are not relevant

**Pull request merged**